### PR TITLE
[UBP] Disable subscriptions to old plans after UBP has been enabled

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

Once UBP has been enabled for a user:

- [x] Disable individual Plan purchases/upgrades (`/plans`)
- [x] Disable old-style Team plan purchases/upgrades (`/teams`)
- [x] Disable new-style Team plan purchases/upgrades (Team Billing) -- upgrade UI gets entirely removed with UBP

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/14940

## How to test

1. Upgrade to a paid Chargebee plan in https://jx-disable-chargebee.preview.gitpod-dev.com/plans
2. Buy a legacy Team Plan in https://jx-disable-chargebee.preview.gitpod-dev.com/teams
3. Once it's active, assign yourself one seat of this Team Plan
4. Now, create a team called "Gitpod [Something]" to unlock usage-based
5. If you now go to https://jx-disable-chargebee.preview.gitpod-dev.com/plans, you can no longer upgrade to any Chargebee paid plan -- you can only cancel by going into `billing` ⏩
6. If you go to https://jx-disable-chargebee.preview.gitpod-dev.com/teams, you can no longer create a new legacy Team Plan -- you can only cancel by going into `billing` ⏩

## Release Notes

```release-note
[gitpod.io] Disable upgrades to fixed-price monthly plans (Chargebee) for individuals and teams who have pay-as-you-go enabled (Stripe)
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by jankeromnes** [comment]
Working well! 🎉

| | /plans | /teams |
| :---: | :---: | :---: |
| **BEFORE UBP** | Screenshot 2022-12-02 at 10 11 51 Can upgrade to individual plans | Screenshot 2022-12-02 at 10 12 00 Can create and manage legacy team plans |
| **AFTER UBP** | Screenshot 2022-12-02 at 10 14 45 Can no longer upgrade to Chargebee plans | Screenshot 2022-12-02 at 10 14 55 Can no longer create legacy team plans |

But, importantly, by visiting the old `/plans` or `/teams` URLs and clicking on billing, you can still view/download your old Chargebee invoices:

Screenshot 2022-12-02 at 10 15 08

**Comment by jankeromnes** [comment]
Hint for reviewers: Reviewing while ignoring indentation changes is easier -- https://github.com/gitpod-io/gitpod/pull/15092/files?w=1

**Comment by svenefftinge** [comment]
After UBP GA do we actually want to allow any kind of plan changes other than cancellation?
We might just allow cancelling

**Comment by jankeromnes** [comment]
>  We might just allow cancelling

This PR disables all individual Plan actions when you're on UBP (i.e. Downgrade, Upgrade). You can still cancel via the Billing interface as usual. I don't think the Billing interface allows any other change than cancel.

**Comment by jankeromnes** [comment]
Hmm, my new change doesn't do exactly what I expected. Back to draft.

**Comment by jldec** [comment]
I tried to downgrade to free after upgrading the UBP while on a personal paid plan.
This seems to be disabled now and the billing link at the top is missing as well
Downgrading to free == cancelling for individual plans, so we should preserve the ability to downgrade to free 

**Before**   (hovering over downgrade to free)
Screenshot 2022-12-02 at 16 48 55

---

**With UBP**  (hovering over downgrade to free)
Screenshot 2022-12-02 at 16 48 17

**Comment by jankeromnes** [comment]
> I tried to downgrade to free after upgrading the UBP while on a personal paid plan.
This seems to be disabled now and ~~the billing link at the top is missing as well~~ (it appeared later, must be a delay)
Downgrading to free == cancelling for individual plans, so we should preserve the ability to downgrade to free 

Many thanks for the feedback @jldec!

The billing link part had me quite worried (should not happen and would break the cancellation UX), but if it came back I agree that it might be some delay or React state bug. In any case, a simple page reload should fix this.

Also, I agree that the Free card action should not get disabled, since downgrading to free is similar to cancelling from 'billing'. Will adjust this.

**Comment by jldec** [comment]
@jankeromnes, when I first opened the team plans page, https://jx-disable-chargebee.preview.gitpod-dev.com/teams in the preview env, I was able to create a new plan, and did not see the blue UBP enabled banner. 

Refreshing the page, then disabled the green button, but the page now appears to be stuck on "payment in progress"

Screenshot 2022-12-05 at 10 35 55

**Comment by jankeromnes** [comment]
> @jankeromnes, when I first opened the team plans page, https://jx-disable-chargebee.preview.gitpod-dev.com/teams in the preview env, I got the dialog to create the new plan, and did not see the blue UBP enabled banner.
> 
> Refreshing the page, then disabled the green button, and showed the banner

Thanks! Then I guess this PR behaves mostly as expected.

The fact that a refresh is needed is likely due to a bug in how the dashboard-side feature flags are implemented (they don't seem to be using a "dispatch", and they don't seem to be eagerly updated in App.tsx or Menu.tsx). In any case, this problem is out-of-scope for this PR. I guess the most important thing is that, once you have UBP enabled, eventually it's no longer possible to upgrade to new Chargebee plans. This seems to be the case.

> and I expect the "payment in progress" message to time out eventually.

It should automatically time out in every browser after a few minutes. If not, that's a but in the Teams page (which is also out of scope here).

**Comment by svenefftinge** [comment]
> Your account has been enabled for usage-based billing

We should say that the current plan is deprecated and will be cancelled by End of March 2023. 
Also talk about the benefits of the new pricing plans (workspace classes and flexible timeouts) would be good before pointing to "how to switch".

**Comment by jankeromnes** [comment]
Back to Draft to implement the reword / redesign suggestion.

**Comment by geropl** [comment]
@jankeromnes Note that we already limit API interactions based on BillingMode. I did not read the whole context on this PR, but wanted to mentioning it because it make sense to keep that in sync with what we do in the UI.

**Comment by jankeromnes** [comment]
> We should say that the current plan is deprecated and will be cancelled by End of March 2023.
Also, talking about the benefits of the new pricing plans (workspace classes and flexible timeouts) would be good before pointing to "how to switch".

@svenefftinge Okay, here is my counter-proposal:

| Plans | Teams |
| --- | --- |
| Screenshot 2022-12-05 at 19 17 48 | Screenshot 2022-12-05 at 19 18 18 |

> The disabled cards are still showing a hover effect that suggests they are clickable. Maybe we should not show those cards at all? Just have a big 'Cancel' button.

I've removed the hover effect by adding the `pointer-events-none` class to disabled Plan cards (this removes all hover and click effects, even if you have listeners).

On the other hand, while I like your more radical proposal of only having a big 'Cancel' button (or, even better, having a new UI that takes users through a single flow from Chargebee cancellation to Stripe re-upgrade), I don't think it's trivial to get the UI right in a short time. It might also be confusing to significantly change the Plans page layout, making it unrecognizable and confusing.

Thus, I'd argue that we merge this PR as is now, and figure out whether a better disabled UI is needed or desirable after GA. 😇

File: components/dashboard/src/settings/Plans.tsx [hunk 1]
```diff
@@ -629,101 +631,99 @@ export default function () {
...
+                    <Alert type="message" className="mb-4">
+                        Your account has been enabled for usage-based billing. Discover faster workspace classes and
+                        only pay for what you actually use.{" "}
+                        <a className="gp-link" href="https://www.gitpod.io/docs/configure/billing/usage-based-billing">
+                            Learn more
+                        </a>
+                        <br />
+                        <br />
+                        The old monthly plans are deprecated and will be cancelled by End of March 2023.
+                    </Alert>
```
Comment (gtsiolis) [inline components/dashboard/src/settings/Plans.tsx:647]: Hey Jan! Some issues on the copy used here:
1. Too verbose, or uses unnecessary empty space.
2. Focus on the personal account, not [team plans](https://www.gitpod.io/docs/configure/billing/team-plans) or [team billing](https://www.gitpod.io/docs/configure/billing/team-billing).
3. Sounds like the user has been part of a closed beta list and not accessing a GA feature on Gitpod.io.
5. Voice and tone is not direct, using terms like _discover_. 

What do you think of the following copy?

| BEFORE | AFTER |
|-|-|
| alert-before | alert-after |

**Review by gtsiolis — COMMENTED** [review]
Left some comments on the copy used in the alert. 🚨

**Comment by jldec** [comment]
@jankeromnes - The notification copy to use is:

> To access [large workspaces](https://www.gitpod.io/docs/configure/workspaces/workspace-classes) and [pay-as-you-go](https://www.gitpod.io/docs/configure/billing/pay-as-you-go), first cancel your existing plan. 
> Existing plans will keep working until the end of March, 2023.

Thanks @gtsiolis  and @svenefftinge  for your inputs.
🙏 

Details behind this wording are in this [RFC](https://www.notion.so/gitpod/RFC-UBP-old-plans-end-of-life-notice-86874db0ec7e4856a1e12344d2f036d8) (internal)

File: components/dashboard/src/settings/Plans.tsx (hunk 1 above)
Comment (svenefftinge) [reply components/dashboard/src/settings/Plans.tsx:647]: Let's not call our pricing as a whole "pay-as-you-go". the pay as you go bit is only the mode where you pay later. We have a free plan and a personal plan with a certain amount of credits that is not pay-as-you-go. We'll likely have other ways to prepurchase credits in the future. 
"Usage-based" is ok but I'd prefer to generally just talk about (new) pricing.

**Review by svenefftinge — APPROVED** [review]
Looks good. I know there is an ongoing discussion about the exact wording, but the current state is good already and we can do the adjustments in even smaller PRs later.

File: components/dashboard/src/settings/Plans.tsx (hunk 1 above)
Comment (gtsiolis) [reply components/dashboard/src/settings/Plans.tsx:647]: DEAL 🤝

**Pull request merged**